vendor/engines/secure_rooms/app/services/secure_rooms/access_handlers/order_handler.rb

```diff
@@ -17,29 +17,33 @@ def initialize(occupancy)
       def process
         return unless occupancy.account_id?
 
-        ActiveRecord::Base.transaction do
-          create_order
-          create_order_detail
-
-          order.validate_order!
-          order.purchase!
-        end
+        create_order unless occupancy.order_detail_id?
+        fulfill_order if occupancy.complete?
 
         order
       end
 
       private
 
       def create_order
+        ActiveRecord::Base.transaction do
+          create_order_and_detail
+          order.validate_order!
+          order.purchase!
+        end
+      end
+
+      def fulfill_order
+        order_detail.update_order_status! occupancy.user, OrderStatus.complete.first
+      end
+
+      def create_order_and_detail
         @order = Order.create!(
           account: occupancy.account,
           user: occupancy.user,
           facility: occupancy.facility,
           created_by_user: occupancy.user,
         )
-      end
-
-      def create_order_detail
         @order_detail = order.order_details.create!(
```

Comment:
There's some almost circular logic here with your instance variables and methods (or at least it's very circuitous to me). I think it might be good to get rid of the memoized order/order_detail below and just pass around the instances. Maybe there's still a better way?